litestar/contrib/sqlalchemy/dto.py

```diff
@@ -218,6 +225,9 @@ def _detect_defaults(elem: ElementType) -> tuple[Any, Any]:
             def default_factory(d: Any = sqla_default) -> Any:
                 return d.arg({})
 
+        elif sqla_default.is_sequence:
+            default = sqla_default
```

Comment:
Does this actually work for you? E.g., in this test:

```py
from litestar.contrib.sqlalchemy.base import BigIntBase
from sqlalchemy import create_engine
from sqlalchemy.orm import Mapped, sessionmaker

from litestar import Litestar, post
from litestar.contrib.sqlalchemy.dto import SQLAlchemyDTO


engine = create_engine("sqlite:///db.sqlite3")
session_factory = sessionmaker(engine, expire_on_commit=False)


class Person(BigIntBase):
    name: Mapped[str]


class PersonDTO(SQLAlchemyDTO[Person]):
    pass


@post("/", dto=PersonDTO)
async def post_person(data: Person) -> Person:
    with session_factory() as session:
        session.add(data)
        session.commit()
    return data


def startup_db() -> None:
    BigIntBase.metadata.create_all(engine)


app = Litestar(route_handlers=[post_person], on_startup=[startup_db])
```
If I make a POST without a value for the id field, such as `{"name": "bob"}` I get the following exception when trying insert the new instance:

```python-traceback
Traceback (most recent call last):
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
sqlite3.ProgrammingError: Error binding parameter 2: type 'Sequence' is not supported

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/peter/PycharmProjects/litestar/litestar/middleware/exceptions/middleware.py", line 157, in __call__
    await self.app(scope, receive, send)
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 77, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 129, in _get_response_for_request
    response = await self._call_handler_function(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 158, in _call_handler_function
    response_data, cleanup_group = await self._get_response_data(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 207, in _get_response_data
    data = await route_handler.fn.value(**parsed_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar-hello-world/app.py", line 25, in post_person
    session.commit()
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1920, in commit
    trans.commit(_to_root=True)
  File "<string>", line 2, in commit
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1236, in commit
    self._prepare_impl()
  File "<string>", line 2, in _prepare_impl
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1211, in _prepare_impl
    self.session.flush()
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4163, in flush
    self._flush(objects)
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4298, in _flush
    with util.safe_reraise():
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4259, in _flush
    flush_context.execute()
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
    rec.execute(self)
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 642, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 93, in save_obj
    _emit_insert_statements(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 1043, in _emit_insert_statements
    result = connection.execute(
             ^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1412, in execute
    return meth(
           ^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 483, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1635, in _execute_clauseelement
    ret = self._execute_context(
          ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1844, in _execute_context
    return self._exec_single_context(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1984, in _exec_single_context
    self._handle_dbapi_exception(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 2339, in _handle_dbapi_exception
    raise sqlalchemy_exception.with_traceback(exc_info[2]) from e
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1965, in _exec_single_context
    self.dialect.do_execute(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/default.py", line 921, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.ProgrammingError: (sqlite3.ProgrammingError) Error binding parameter 2: type 'Sequence' is not supported
[SQL: INSERT INTO person (name, id) VALUES (?, ?)]
[parameters: ('bob', Sequence('person_id_seq', metadata=MetaData()))]
(Background on this error at: https://sqlalche.me/e/20/f405)
```

Reply:
It was working fine for me as of yesterday. I was using the repository methods but don't see that that should make a difference. But it sounds like you're going a different route anyway.

Reply:
> But it sounds like you're going a different route anyway.

No, I'm not working on anything else related to support for sequence types. The other PR is for type inference where a model doesn't have a type hint for a column, which I don't believe is related to the core issue of #1851.

Reply:
Tried changing that line to
`default_factory = sqla_default.next_value`
Get error:
`NotImplementedError: Dialect 'sqlite' does not support sequence increments.`
Might be something else to ask in SQLAlchemy tomorrow.

Reply:
Yeah, I don't think we should be trying to extract a value from the sequence here. My understanding is that they represent a sequence on the database side, and are ignored if the database backend doesn't support sequences.

On that thought, this should probably just be:

```py
if sqla_default.is_sequence:
    continue
```

Reply:
I realized why it was working in my case; I was using DTOData rather than passing directly.

I tried changing to
```
elif sqla_default.is_sequence:
    return Empty, None
```
but then I get "Object missing required field `id`." The DTO is still going to want to set a default from somewhere.

Btw, thanks for your help with this; I'm new to the guts of litestar and sqlalchemy so a lot of this is unfamiliar territory, just trying to be helpful since this was blocking me on a personal project.

Reply:
Yes, and that is exactly what `DTOData` is for - when the client is not able to pass complete enough information to instantiate an instance of the model type.

Some relevant docs:
- https://docs.litestar.dev/2/usage/dto/1-dto-factory.html#dto-data
- https://docs.litestar.dev/2/tutorials/dto-tutorial/07-read-only-fields.html#read-only-fields

Reply:
Oh yeh, `continue` was wrong, should be:

```py
elif sqla_default.is_sequence:
    pass
```
That will leave the default as `Empty`.

Reply:
Fair enough, but the example you gave would work fine if it inherited from one of the UUID base classes. And it seems like since the database handles ID creation and it's such a common use case. this would be a pain point for new users.

Reply:
It does work on SQLAlchemy if I change it to:

```
elif sqla_default.is_sequence:
    default = None
```

Reply:
> Fair enough, but the example you gave would work fine if it inherited from one of the UUID base classes.

Yes, but the UUID PK columns have a default that is generated in python by sqlalchemy, not by the database server - this is a key difference between the two:

https://github.com/litestar-org/litestar/blob/48a9719b1ad5e94b12f1b6755381d461cf517f4d/litestar/contrib/sqlalchemy/base.py#L91

> And it seems like since the database handles ID creation and it's such a common use case, this would be a pain point for new users.

This is _exactly_ why `DTOData` exists, as I've explained. If you want to be able to have a `None` default value for your ID column then you can type it `int | None` and give it a default value, but that's up to you. You also do not need to use the DTOs at all.

> It does work with sqlite if I change it to:

You are setting a default value for the column of `None` here - see comment above about `DTOData`. In most cases you don't want a `None` default value, you want a non-supplied value - very different things.

Reply:
Also, are you using SQLAlchemy with the `MappedAsDataclass` mixin? Without that, sqlalchemy doesn't require values for all required columns at model instantiation time, and so in that case you don't need `DTOData`.

Reply:
>This is exactly why DTOData exists, as I've explained. If you want to be able to have a None default value for your ID column then you can type it int | None and give it a default value, but that's up to you.

OK, fair enough. I would at least suggest adding a specific note about this in the tutorial and documentation, as it seems strange that your two provided SQLAlchemy bases would have different behavior, and not obvious without digging through the library source code.

> You also do not need to use the DTOs at all.

For such a basic use case as this, that seems to defeat a large part of the purpose of using `litestar`.

> Also, are you using SQLAlchemy with the MappedAsDataclass mixin?

No, I'm using the builtin `BigIntAuditBase` class.

Anyway, if this is the route you want to go, then I can withdraw the PR.

Reply:
@peterschutt I pushed a commit which I think does what you are suggesting. I don't know that I'm entirely satisfied with this resolution, but at least it unblocks things for now.

Reply:
Sorry, but this is still not right.

This is the exception that your current revision yields with the same test case as I posted above:

```python-traceback
Traceback (most recent call last):
  File "/home/peter/PycharmProjects/litestar/litestar/middleware/exceptions/middleware.py", line 157, in __call__
    await self.app(scope, receive, send)
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 77, in handle
    response = await self._get_response_for_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 129, in _get_response_for_request
    response = await self._call_handler_function(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 158, in _call_handler_function
    response_data, cleanup_group = await self._get_response_data(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar/litestar/routes/http.py", line 207, in _get_response_data
    data = await route_handler.fn.value(**parsed_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/PycharmProjects/litestar-hello-world/app.py", line 26, in post_person
    session.commit()
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1920, in commit
    trans.commit(_to_root=True)
  File "<string>", line 2, in commit
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1236, in commit
    self._prepare_impl()
  File "<string>", line 2, in _prepare_impl
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/state_changes.py", line 139, in _go
    ret_value = fn(self, *arg, **kw)
                ^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 1211, in _prepare_impl
    self.session.flush()
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4163, in flush
    self._flush(objects)
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4298, in _flush
    with util.safe_reraise():
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py", line 147, in __exit__
    raise exc_value.with_traceback(exc_tb)
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/session.py", line 4259, in _flush
    flush_context.execute()
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 466, in execute
    rec.execute(self)
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/unitofwork.py", line 642, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 93, in save_obj
    _emit_insert_statements(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/orm/persistence.py", line 1220, in _emit_insert_statements
    result = connection.execute(
             ^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1412, in execute
    return meth(
           ^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 483, in _execute_on_connection
    return connection._execute_clauseelement(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/engine/base.py", line 1627, in _execute_clauseelement
    compiled_sql, extracted_params, cache_hit = elem._compile_w_cache(
                                                ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 671, in _compile_w_cache
    compiled_sql = self._compiler(
                   ^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/elements.py", line 288, in _compiler
    return dialect.statement_compiler(dialect, self, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 1425, in __init__
    Compiled.__init__(self, dialect, statement, **kwargs)
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 866, in __init__
    self.string = self.process(self.statement, **compile_kwargs)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 911, in process
    return obj._compiler_dispatch(self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
           ^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 5626, in visit_insert
    crud_params_struct = crud._get_crud_params(
                         ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/crud.py", line 314, in _get_crud_params
    use_insertmanyvalues, use_sentinel_columns = _scan_cols(
                                                 ^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/crud.py", line 658, in _scan_cols
    _append_param_parameter(
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/crud.py", line 877, in _append_param_parameter
    value = compiler.process(
            ^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 911, in process
    return obj._compiler_dispatch(self, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/visitors.py", line 143, in _compiler_dispatch
    return meth(self, **kw)  # type: ignore  # noqa: E501
           ^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 2865, in visit_function
    text = disp(func, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 2898, in visit_next_value_func
    return self.visit_sequence(next_value.sequence)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/peter/.cache/pypoetry/virtualenvs/litestar-hello-world-bK1jcEeo-py3.11/lib/python3.11/site-packages/sqlalchemy/sql/compiler.py", line 2901, in visit_sequence
    raise NotImplementedError(
NotImplementedError: Dialect 'sqlite' does not support sequence increments.
```

Just FYI statements such as:

> For such a basic use case as this, that seems to defeat a large part of the purpose of using litestar.

and 

> Anyway, if this is the route you want to go...

... come across as inflammatory - the only route I want to go, is the correct one. But I'll give the benefit of the doubt and try to explain why I cannot accept this PR as it stands.

```py

        elif sqla_default.is_sequence:

            def default_factory(d: Any = sqla_default) -> Any:
                return d.next_value()
```
In this code, if a value is not submitted by the client for the `id` field, this sets the value of the `id` field on the instance that we create to inject into the handler to the result of the `next_value()` call.

[`Sequence.next_value()`](https://docs.sqlalchemy.org/en/20/core/defaults.html#sqlalchemy.schema.Sequence.next_value) returns a [`Function`](https://docs.sqlalchemy.org/en/20/core/functions.html#sqlalchemy.sql.functions.Function) object, but the type of `BigIntBase.id` is `Mapped[int]`, not `Mapped[Function[int]]`. So what you are proposing is not type safe. This is the same as why we cannot set the default to `None`, the type of `BigIntBase.id` is not `Mapped[int | None]`, so setting its value to `None` is not type safe either.

What do I mean by "not type safe"?

If a user tries to perform an integer operation on the value of the `id` field after we inject it, and its value is not an integer, then we've broken a contract. This code, for example:

```py
@post("/", dto=PersonDTO)
async def post_person(data: Person) -> Person:
    if data.id > 123:
        raise ValueError("Invalid ID")
    with session_factory() as session:
        session.add(data)
        session.commit()
    return data
```
... will fail with an error like:

> `TypeError: Boolean value of this clause is not defined`

Because the `id` field is a `sqlalchemy.sql.Function` instance, not an integer.

> OK, fair enough. I would at least suggest adding a specific note about this in the tutorial and documentation, as it seems strange that your two provided SQLAlchemy bases would have different behavior, and not obvious without digging through the library source code.

Please open new issues for any documentation changes you'd like to see.

@cofin what are your thoughts on the differing behavior between the base model types WRT to primary key defaults?